packages/mcfunction/src/parser/macro.ts

```diff
@@ -0,0 +1,103 @@
+import * as core from '@spyglassmc/core'
+import { localeQuote, localize } from '@spyglassmc/locales'
+import type { MacroNode } from '../node/index.js'
+
+/**
+ * Parse a macro line.
+ */
+export function macro(): core.Parser<MacroNode> {
+	return (
+		src: core.Source,
+		ctx: core.ParserContext,
+	): core.Result<MacroNode> => {
+		const ans: MacroNode = {
+			type: 'mcfunction:macro',
+			range: core.Range.create(src.cursor),
+			children: [],
+		}
+
+		// Skip the starting '$'
+		src.skip()
+
+		let start = src.cursor
+		let hasMacroArgs = false
+
+		while (src.canReadInLine()) {
+			src.skipUntilOrEnd(core.LF, core.CR, '$')
+			if (src.peek(2) === '$(') {
+				// Add the other stuff before this macro key
+				const other = src.sliceToCursor(start)
+				if (other.length > 0) {
+					ans.children.push({
+						type: 'mcfunction:macro/other',
+						range: core.Range.create(start, src.cursor),
+						value: other,
+					})
+					start = src.cursor
+				}
+				// Parse the macro key
+				src.skip(2)
+				const keyStart = src.cursor
+				src.skipUntilOrEnd(core.LF, core.CR, ')')
+				if (src.peek() != ')') {
+					// Macro key was not closed
+					ctx.err.report(
+						localize('expected', localeQuote(')')),
+						core.Range.create(keyStart, src.cursor),
+					)
+				} else if (src.cursor <= keyStart) {
+					// Encountered $()
+					ctx.err.report(
+						localize('expected', localize('mcfunction.parser.macro.key')),
+						core.Range.create(start, src.cursor + 1),
+					)
+				}
+				const key = src.sliceToCursor(keyStart)
+				const matchedInvalid = key.replace(/[a-zA-Z0-9_]*/, '')
+				if (matchedInvalid.length > 0) {
+					ctx.err.report(
+						localize(
+							'mcfunction.parser.macro.illegal',
+							matchedInvalid.charAt(0),
+						),
+						core.Range.create(keyStart, src.cursor),
+					)
+				}
```

Comment:
can we factor this code section into a helper function? purely for readability -- something like `validateMacro()`

Reply:
its very long and looks like it exclusively checks for invalid macros, it being refactored into a helper would make it clear that everything _else_ in the `macro` function is for actual parsing

Reply:
`validateMacroArgument()` would match the terminology we use here better.